### PR TITLE
SCT Out of Bounds Fix

### PR DESCRIPTION
**Pull request by christopher-henderson**
Addresses #228 @dandragona-dev I admit I'm having a moderately difficult time putting together test data for this one. Do you happen to have an example cert that caused this?

### CODE REVIEW

**Comment by cpu** [comment]
> I admit I'm having a moderately difficult time putting together test data for this one.

@christopher-henderson Does this small test program I put together for making some ZLint test data help at all?  https://gist.github.com/cpu/6d26b2718f29e184ff88a90f02d7cbcb It's a fairly complete example of generating certs with custom SCT lists.

**Review by zakird — APPROVED** [review]
Overall this looks reasonable, though it would be good to get a test in place, if @cpu's proposal is easy to make happen. It looks like there are also some formatting issues that the tests are hung up on.

**Comment by christopher-henderson** [comment]
Sorry, had to get back to this this weekend.

@cpu's script works like a charm for generating certs with variable length SCT lists, however that is not what the issue here was. 

The array in question here is a list of `struct {length_header length_header data}` and the problem was if you get a malformed struct at the end of the list (say, `struct {length_header}` and that's it).

So I'll probably have to abstract this out to a function that returns a `[][]byte` so that I can test it without having to handcraft malformed ASN1.

**Comment by cpu** [comment]
> The array in question here is a list of struct {length_header length_header data} and the problem was if you get a malformed struct at the end of the list (say, struct {length_header} and that's it).

Ahhh indeed, not the same situation. So much for that thought.

> So I'll probably have to abstract this out to a function that returns a [][]byte so that I can test it without having to handcraft malformed ASN1.

If you do go down the road of artisanal ASN1 [der-ascii](https://github.com/google/der-ascii) might make it more palatable. 

I do think this should have test coverage if possible, either through refactor or targetted testcase data.

**Comment by zakird** [comment]
hey @christopher-henderson , just wanted to follow up on this PR before we get further astray from master. Were you able to come up with any tests? It seems like the PR is also failing tests, which we should figure out.

**Comment by christopher-henderson** [comment]
I admit that this got triaged pretty far down in my list (it seemed like it was an observation rather than a crash report). I can give the sample data another go round if there is anything particularly time sensitive here.

The failed tests are just because apparently the changes weren't `gofmted` before pushing (I sometimes forget that GoLand needs to you to set that up per-project).

**Comment by zakird** [comment]
Totally fair! Back story: we have someone who is going to be working on backporting upstream changes to crypto/tls and crypto/x509 to this library in the coming weeks. This will undoubtedly break every pending PR to this repo. This one looks like it's very close (and I'd even be OK merging without a test if we're confident in the code) so was hoping we could get it merged instead of closing it out.

**Comment by cpu** [comment]
I have some spare time today, let me see if I can figure out a test case for this. If I can't then we can merge as-is.

**Comment by cpu** [comment]
> I have some spare time today, let me see if I can figure out a test case for this. If I can't then we can merge as-is.

Got one :tada:

```
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
github.com/zmap/zcrypto/x509.parseCertificate(0xc0002de000, 0x1d1, 0x1d7, 0x5f85a0)
	/home/daniel/go/src/github.com/zmap/zcrypto/x509/x509.go:1914 +0x798f
github.com/zmap/zcrypto/x509.ParseCertificate(0xc0002dc000, 0x1d1, 0x1d7, 0xc00007ef00, 0xc0002da2ac, 0x0)
	/home/daniel/go/src/github.com/zmap/zcrypto/x509/x509.go:1982 +0xf3
main.main()
	/home/daniel/go/src/github.com/zmap/zcrypto/repro.go:15 +0x85
exit status 2
```

Lemme see about cleaning it up and then I'll try to push to this branch.

**Comment by cpu** [comment]
> Lemme see about cleaning it up and then I'll try to push to this branch.

That took me a little bit longer than expected. I flailed around for a while thinking about the best way to test the **success** case where a SCT list is populated. Ultimately I gave up because ZCrypto's fork of `certificate-transparency-go` is missing the types/serialization code to serialize a SCT list. Adding that code just for this test felt silly and using a hard-coded pile of ASN.1 bytes that matched up to an in-memory structure was also yucky. If someone else wants to take a crack at this part be my guest :-)

I was able to add pretty decent coverage to @christopher-henderson's code for the **error** cases in 8d34eed7529644a2e1f168ffab419f21e966e011 and I think that's most important. I also verified that this branch fixes the panic I was able to repro with master based on the description from https://github.com/zmap/zcrypto/issues/228, I think that's sufficient to be confident in merging this branch. 

You can repro a panic in the the x509 code prior to this commit with the following cert:
```
-----BEGIN CERTIFICATE-----
MIIBPjCB6aADAgECAgIFOTANBgkqhkiG9w0BAQsFADAZMRcwFQYDVQQDEw5aQ3J5
cHRvLVBSLTIzNzAiGA8wMDAxMDEwMTAwMDAwMFoYDzAwMDEwMTAxMDAwMDAwWjAZ
MRcwFQYDVQQDEw5aQ3J5cHRvLVBSLTIzNzBcMA0GCSqGSIb3DQEBAQUAA0sAMEgC
QQDkuzu8wTvGkm/dGfN/IYbAeUp1hkOXb7uL+vhkX0UYfU+jKhz2Z4MbfAAe6mj4
OabIKXursXD2pfWiC81Lh9NZAgMBAAGjFzAVMBMGCisGAQQB1nkCBAIEBQQDAAAA
MA0GCSqGSIb3DQEBCwUAA0EAA7/g2nYDNjAA/aJFv/vpT0JThUzrVChgBzlb5v4N
BDRdVTaD1LHZw1KV12U5Q5iqUX6tSmJmKvF1pGiPhLvT0Q==
-----END CERTIFICATE-----
```

e.g. using `zcertificate` with `golang.org/x/crypto v0.0.0-20201208171446-5f87f3452ae9`:

```
> ~/go/bin/zcertificate ./test.pem
INFO[0000] reading from ./test.pem
INFO[0000] writing to stdout
panic: runtime error: index out of range [1] with length 1

goroutine 21 [running]:
github.com/zmap/zcrypto/x509.parseCertificate(0xc00007e000, 0x142, 0x147, 0x665de0)
	/home/daniel/go/pkg/mod/github.com/zmap/zcrypto@v0.0.0-20201211161100-e54a5822fb7e/x509/x509.go:1914 +0x792f
github.com/zmap/zcrypto/x509.ParseCertificate(0xc00002e2c0, 0x142, 0x147, 0x0, 0x0, 0x0)
	/home/daniel/go/pkg/mod/github.com/zmap/zcrypto@v0.0.0-20201211161100-e54a5822fb7e/x509/x509.go:1982 +0xf3
main.processCertificate(0xc0000be4e0, 0xc0000be5a0, 0xc0000b9720)
	/home/daniel/go/src/github.com/zmap/zcertificate/cmd/zcertificate/main.go:68 +0xf7
created by main.main
	/home/daniel/go/src/github.com/zmap/zcertificate/cmd/zcertificate/main.go:170 +0x3f0
```

Updating `zcertificate`'s `go.mod` with a `replace` to use this ZCrypto branch results in a descriptive error instead of a panic:

```
> ~/go/bin/zcertificate ./test.pem
INFO[0000] reading from ./test.pem
INFO[0000] writing to stdout
ERRO[0000] could not parse certificate: malformed SCT extension: trailing data
```

**Comment by cpu** [comment]
@zakird @christopher-henderson Do you folks want to double check my additions and then we can merge? Thanks!

File: x509/x509.go [hunk 1]
```diff
@@ -2025,7 +2025,7 @@ func parseSignedCertificateTimestampList(out *Certificate, ext pkix.Extension) e
 	}
 	scts = scts[2:]
 	headerLength := 2
-	for  {
```
Comment (cpu) [inline x509/x509.go]: this was a fix from `gofmt`

**Review by zakird — APPROVED** [review]
👍🏼

File: x509/x509_test.go [hunk 2]
```diff
@@ -278,17 +275,6 @@ func TestCreateSelfSignedCertificate(t *testing.T) {
...
 
-	ed25519Pub, ed25519Priv, err := ed25519.GenerateKey(nil)
-	if err != nil {
-		t.Fatalf("Failed to generate Ed25519 key: %s", err)
-	}
-	var pubkey, privkey [32]byte
-	if _, err := io.ReadFull(rand.Reader, privkey[:]); err != nil {
-		panic(err)
-	}
-	curve25519.ScalarBaseMult(&pubkey, &privkey)
```
Comment (cpu) [inline x509/x509_test.go:289]: oops, I think I'm undoing some work here.

**Review by cpu — CHANGES_REQUESTED** [review]
Noticed I messed up a merge, fixing.

File: x509/x509_test.go (hunk 2 above)
Comment (cpu) [reply x509/x509_test.go]: Fixed with my force-push to 139f80d

**Review by cpu — APPROVED** [review]

**Comment by cpu** [comment]
Thanks everyone! Team effort :tada:

**Pull request merged**